File: src/main/java/io/koju/autopos/trade/purchase/domain/Purchase.java
package io.koju.autopos.trade.purchase.domain;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonView;
import io.koju.autopos.kernel.json.Views;
import io.koju.autopos.party.domain.Vendor;
import io.koju.autopos.trade.domain.Trade;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import static javax.persistence.GenerationType.SEQUENCE;

@Entity
@Table(name = "purchase")
@Getter
@Setter
public class Purchase extends Trade {

    private static final String ID_SEQ = "purchase_id_seq";

    @Id
    @SequenceGenerator(name = ID_SEQ, sequenceName = ID_SEQ, allocationSize = 1)
    @GeneratedValue(strategy = SEQUENCE, generator = ID_SEQ)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "vendor_id")
    @JsonView(Views.Summary.class)
    private Vendor vendor;

    @OneToMany(mappedBy = "purchase", fetch = FetchType.EAGER)
    @JsonManagedReference
    private List<PurchaseLine> lines = new ArrayList<>();

    @JsonView(Views.Summary.class)
    public BigDecimal getGrandTotal() {
        return lines.stream()
                    .map(PurchaseLine::getAmount)
                    .reduce(BigDecimal.ZERO, BigDecimal::add)
                    .subtract(getDiscount());
    }

}

File: src/main/scala/io/koju/autopos/trade/purchase/domain/Purchase.scala
package io.koju.autopos.trade.purchase.domain

import java.lang.Long
import java.math.BigDecimal
import java.util
import javax.persistence.GenerationType.SEQUENCE
import javax.persistence._

import com.fasterxml.jackson.annotation.{JsonManagedReference, JsonView}
import io.koju.autopos.kernel.json.Views
import io.koju.autopos.party.domain.Vendor
import io.koju.autopos.trade.domain.Trade

import scala.beans.BeanProperty
import scala.collection.JavaConverters._


@Entity
@Table(name = "purchase")
class Purchase extends Trade {

  @Id
  @SequenceGenerator(name = "purchase_id_seq", sequenceName = "purchase_id_seq", allocationSize = 1)
  @GeneratedValue(strategy = SEQUENCE, generator = "purchase_id_seq")
  @BeanProperty
  var id: Long = _

  @ManyToOne
  @JoinColumn(name = "vendor_id")
  @JsonView(Array(classOf[Views.Summary]))
  @BeanProperty
  var vendor: Vendor = _

  @OneToMany(mappedBy = "purchase", fetch = FetchType.EAGER)
  @JsonManagedReference
  @BeanProperty
  var lines: util.List[PurchaseLine] = new util.ArrayList[PurchaseLine]

  @JsonView(Array(classOf[Views.Summary]))
  def getGrandTotal: BigDecimal = {
    lines.asScala.map(_.getAmount).reduceOption(_.add(_)).map(_.subtract(getDiscount)).getOrElse(BigDecimal.ZERO)
  }

}